### DISCUSSION TITLE
Handling of breaking changes

### DISCUSSION THREAD

**Discussion by smjonas** [kylechui/nvim-surround#108] [General]
I noticed there were quite a few breaking changes recently and issues because of that. Maybe it might be worth using semantic versioning for your plugin. Then users can avoid dealing with breaking changes during their normal workflow.

Just putting this out here for general discussion / comments. I could help you with a script that automatically pushes a new version tag to GitHub when you increase the version number.

**Comment by kylechui** [comment]
I think at this point in time the plugin could *maybe* warrant some versioning system, although I don't really have any experience with that. I was thinking of calling post #60, #101 `v0.1`, and figuring it out from there. I'm open to any PRs or more concrete ideas you might have though :)

**Reply by akinsho** [reply]
@kylechui I've also been thinking about this as I've started to add more buffer type configs and having to refactor options over time will be a pain. One thing I think is quite good to do is to do soft deprecations where possible. If feasible e.g. you move a field to another table then you can make use of lua metatables with something like an `__index` function e.g. say you moved `config.delimiters` to `config.nested.delimiters`

you could do
```lua
setmetatable(config, {
  __index = function(tbl, key)
    if key == "delimiter" then
      vim.notify_once("delimiter has been moved to 'nested.delimiter'")
      return tbl.nested.delimiter[key]
    end
  end
})
```

this is just one such idea it might not always be best to keep multiple implementations around but I think if the burden is low or you can easily add a temporary workaround to make peoples lives easier then that would be good.

**Reply by kylechui** [reply]
Oh wow, metatables certainly seem quite useful. I think I'll consider using them in the future after #113, but as of right now that pull request introduces immense restructuring of the internals, including user configuration. My goal at the moment is to "get it mostly right" for the upcoming changes, and try out your suggestion for minor patches in the future. Thanks for the tip!

**Reply by kylechui** [reply]
@smjonas @akinsho @andrewferrier I think the most recent commit addresses nearly everything I have had an issue with; the two things I've found that I *might* fix/add in the future are:
* Modifying lines with an *odd* number of quotes is weird for the final quote, e.g. `ds"` on the last column of `"three"quotes"`
* Deletions/changes can't take user input, at least not well (dot-repeating is also finicky)
  * I think this can be delayed for a later release, as there are other more "pressing" issues that I should probably focus on

I'm currently thinking of releasing this under `v1.0.0`; to my understanding the major version update is to signify major public API changes, the minor for adding smaller features (may include smaller API changes), and patches for bug fixes. In the future the first bullet above could be a patch, the second a minor update. If any of y'all have questions on configuration, please read the help docs; otherwise tag me with any points of confusion. I'll sit on this for the next few days, and if no major bugs are found then I'll merge into main :tada:

**Reply by smjonas** [reply]
@kylechui Really amazing work from you on the `pattern-matching` branch the last weeks, impressive! I just left you some comments on the PR.
